File: java/src/main/java/io/github/xesam/linkseal/resolver/TemplateEngine.java
package io.github.xesam.linkseal.resolver;

import java.nio.charset.StandardCharsets;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.StringJoiner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Expands the LinkSeal-supported subset of RFC 6570 URI templates.
 *
 * <p>Supported expressions (and only these):
 * <ul>
 *   <li>{@code {var}} — simple substitution (Level 1), one or more comma-separated names</li>
 *   <li>{@code {?var}} — form-style query start</li>
 *   <li>{@code {&amp;var}} — form-style query continuation</li>
 * </ul>
 *
 * <p>Unsupported RFC 6570 operators ({@code + # . / ;}) and modifiers (explode {@code *},
 * prefix {@code :N}) are <strong>rejected</strong> rather than silently mis-expanded, so this
 * SDK accepts and rejects the same template grammar as the JavaScript SDK. Variable names are
 * restricted to {@code [A-Za-z0-9_.]}.
 *
 * <p>Value encoding follows strict RFC 6570, which defers to RFC 3986 "unreserved":
 * {@code A-Za-z0-9-._~} pass through unchanged; every other byte is percent-encoded as its
 * UTF-8 octet. This is hand-rolled (not delegated to a library) so both SDKs emit byte-identical
 * output regardless of which URI-charset generation a library froze at.
 */
public final class TemplateEngine {
  /** Matches a supported expression: optional {@code ?}/{@code &} operator + comma-separated names. */
  private static final Pattern SUPPORTED =
      Pattern.compile("\\{([?&]?)([A-Za-z0-9_.]+(?:,[A-Za-z0-9_.]+)*)\\}");

  /** Matches any braced expression so unsupported ones can be flagged with a precise message. */
  private static final Pattern ANY_BRACE = Pattern.compile("\\{([^}]*)\\}");

  private static final String OPERATORS = "+#./;?&";
  private static final Pattern VARNAME = Pattern.compile("[A-Za-z0-9_.]+");

  public String expand(String template, Map<String, String> variables) {
    validate(template);

    Matcher matcher = SUPPORTED.matcher(template);
    StringBuffer out = new StringBuffer();
    while (matcher.find()) {
      String operator = matcher.group(1);
      String expression = matcher.group(2);
      String replacement = switch (operator) {
        case "?" -> expandQuery("?", expression, variables);
        case "&" -> expandQuery("&", expression, variables);
        default -> expandSimple(expression, variables);
      };
      matcher.appendReplacement(out, Matcher.quoteReplacement(replacement));
    }
    matcher.appendTail(out);
    return out.toString();
  }

  /** Placeholder names declared by the template — the basis for missing-variable detection. */
  public List<String> variableNames(String template) {
    validate(template);

    Matcher matcher = SUPPORTED.matcher(template);
    Set<String> names = new LinkedHashSet<>();
    while (matcher.find()) {
      for (String name : matcher.group(2).split(",")) {
        names.add(name);
      }
    }
    return List.copyOf(names);
  }

  private String expandSimple(String expression, Map<String, String> variables) {
    StringJoiner joiner = new StringJoiner(",");
    for (String name : expression.split(",")) {
      joiner.add(encode(variables.getOrDefault(name, "")));
    }
    return joiner.toString();
  }

  private String expandQuery(String prefix, String expression, Map<String, String> variables) {
    StringJoiner joiner = new StringJoiner("&");
    for (String name : expression.split(",")) {
      String value = variables.get(name);
      if (value != null) {
        joiner.add(encode(name) + "=" + encode(value));
      }
    }

    String query = joiner.toString();
    return query.isEmpty() ? "" : prefix + query;
  }

  /**
   * Rejects unsupported RFC 6570 operators and modifiers so both SDKs accept and reject the same
   * template grammar. A braced expression that isn't a supported form throws here, before any
   * expansion can produce a URL that differs from the JavaScript SDK.
   */
  private void validate(String template) {
    Matcher matcher = ANY_BRACE.matcher(template);
    while (matcher.find()) {
      String expr = matcher.group(1);
      String operator = "";
      int bodyStart = 0;
      if (!expr.isEmpty() && OPERATORS.indexOf(expr.charAt(0)) >= 0) {
        operator = expr.substring(0, 1);
        bodyStart = 1;
      }
      if (!operator.isEmpty() && !operator.equals("?") && !operator.equals("&")) {
        throw new IllegalArgumentException(
            "Unsupported RFC 6570 operator '{" + operator + "}' in template: " + template
                + " — LinkSeal supports only {}, {?}, and {&}");
      }
      for (String spec : expr.substring(bodyStart).split(",")) {
        String name = spec.trim();
        if (name.isEmpty()) {
          throw new IllegalArgumentException("Empty variable spec in template: " + template);
        }
        if (name.indexOf('*') >= 0 || name.indexOf(':') >= 0) {
          throw new IllegalArgumentException(
              "Unsupported modifier in '" + name + "': explode (*) and prefix (:N) are not supported");
        }
        if (!VARNAME.matcher(name).matches()) {
          throw new IllegalArgumentException(
              "Invalid variable name '" + name + "' in template: " + template);
        }
      }
    }
  }

  private static final char[] HEX = "0123456789ABCDEF".toCharArray();

  /** RFC 3986 unreserved (A-Za-z0-9-._~) passes through; all other bytes are %XX UTF-8. */
  private String encode(String value) {
    byte[] bytes = value.getBytes(StandardCharsets.UTF_8);
    StringBuilder out = new StringBuilder(bytes.length);
    for (byte b : bytes) {
      int u = b & 0xFF;
      if ((u >= 0x30 && u <= 0x39) // 0-9
          || (u >= 0x41 && u <= 0x5A) // A-Z
          || (u >= 0x61 && u <= 0x7A) // a-z
          || u == 0x2D // -
          || u == 0x2E // .
          || u == 0x5F // _
          || u == 0x7E // ~
      ) {
        out.append((char) u);
      } else {
        out.append('%').append(HEX[u >>> 4]).append(HEX[u & 0x0F]);
      }
    }
    return out.toString();
  }
}

File: javascript/packages/unilink-seal-resolver/src/TemplateEngine.ts
/**
 * Expands the LinkSeal-supported subset of RFC 6570 URI templates.
 *
 * Supported expressions (and only these):
 *   {var}   — simple substitution (Level 1), one or more comma-separated names
 *   {?var}  — form-style query start
 *   {&var}  — form-style query continuation
 *
 * Unsupported RFC 6570 operators (+ # . / ;) and modifiers (explode *, prefix :N) are
 * rejected rather than silently mis-expanded, so this SDK accepts and rejects the exact
 * same template grammar as the Java SDK. Variable names are restricted to [A-Za-z0-9_.].
 *
 * Value encoding follows strict RFC 6570, which defers to RFC 3986 "unreserved":
 * the characters A-Za-z0-9-._~ pass through unchanged; every other byte is percent-encoded
 * as its UTF-8 octet. This is hand-rolled (rather than delegating to a library) so both
 * SDKs emit byte-identical output regardless of which URI-charset generation a library
 * froze at — see spec/PROTOCOL.md §4.4 and §5.
 */
import { utf8Encode } from './shared/utf8.js';
import { ResolutionError } from './errors.js';

const ANY_BRACE = /\{([^}]*)\}/g;
const OPERATORS = '+#./;?&';
const VARNAME = /^[A-Za-z0-9_.]+$/;

/** Unbalanced or nested braces make the template invalid — reject rather than emit them
 *  as literals. Both SDKs run this before expression-level validation. */
function assertBracesBalanced(template: string): void {
  let depth = 0;
  for (const ch of template) {
    if (ch === '{') {
      if (depth > 0) {
        throw new ResolutionError('INVALID_TEMPLATE', `Nested '{' in template: ${template}`);
      }
      depth++;
    } else if (ch === '}') {
      if (depth === 0) {
        throw new ResolutionError('INVALID_TEMPLATE', `Unmatched '}' in template: ${template}`);
      }
      depth--;
    }
  }
  if (depth > 0) {
    throw new ResolutionError('INVALID_TEMPLATE', `Unclosed '{' in template: ${template}`);
  }
}

function validateTemplate(template: string): void {
  assertBracesBalanced(template);
  for (const match of template.matchAll(ANY_BRACE)) {
    const expr = match[1];
    let operator = '';
    let bodyStart = 0;
    if (expr.length > 0 && OPERATORS.includes(expr[0])) {
      operator = expr[0];
      bodyStart = 1;
    }
    if (operator !== '' && operator !== '?' && operator !== '&') {
      throw new ResolutionError(
        'UNSUPPORTED_TEMPLATE',
        `Unsupported RFC 6570 operator '{${operator}}' in template: ${template} — LinkSeal supports only {}, {?}, and {&}`
      );
    }
    for (const spec of expr.slice(bodyStart).split(',')) {
      const name = spec;
      if (name === '') {
        throw new ResolutionError('INVALID_TEMPLATE', `Empty variable spec in template: ${template}`);
      }
      if (name.includes('*') || name.includes(':')) {
        throw new ResolutionError(
          'UNSUPPORTED_TEMPLATE',
          `Unsupported modifier in '${name}': explode (*) and prefix (:N) are not supported`
        );
      }
      if (!VARNAME.test(name)) {
        throw new ResolutionError(
          'INVALID_TEMPLATE',
          `Invalid variable name '${name}' in template: ${template}`
        );
      }
    }
  }
}

/** RFC 3986 unreserved (A-Za-z0-9-._~) passes through; all other bytes are %XX UTF-8. */
function encodeValue(value: string): string {
  let out = '';
  for (const byte of utf8Encode(value)) {
    if (
      (byte >= 0x30 && byte <= 0x39) || // 0-9
      (byte >= 0x41 && byte <= 0x5a) || // A-Z
      (byte >= 0x61 && byte <= 0x7a) || // a-z
      byte === 0x2d || // -
      byte === 0x2e || // .
      byte === 0x5f || // _
      byte === 0x7e    // ~
    ) {
      out += String.fromCharCode(byte);
    } else {
      out += '%' + byte.toString(16).toUpperCase().padStart(2, '0');
    }
  }
  return out;
}

export class TemplateEngine {
  expand(template: string, variables: Record<string, string>): string {
    validateTemplate(template);
    let out = '';
    let last = 0;
    for (const match of template.matchAll(ANY_BRACE)) {
      const index = match.index ?? 0;
      out += template.slice(last, index);
      last = index + match[0].length;

      const expr = match[1];
      let operator = '';
      let body = expr;
      if (expr.length > 0 && OPERATORS.includes(expr[0])) {
        operator = expr[0];
        body = expr.slice(1);
      }
      const names = body.split(',');

      if (operator === '?' || operator === '&') {
        const pairs = names
          .filter((name) => variables[name] !== undefined && variables[name] !== null)
          .map((name) => `${name}=${encodeValue(variables[name])}`);
        out += pairs.length ? operator + pairs.join('&') : '';
      } else {
        out += names.map((name) => encodeValue(variables[name] ?? '')).join(',');
      }
    }
    return out + template.slice(last);
  }

  /** Placeholder names declared by the template — the basis for missing-variable detection. */
  variableNames(template: string): string[] {
    validateTemplate(template);
    const names = new Set<string>();
    for (const match of template.matchAll(ANY_BRACE)) {
      const expr = match[1];
      let body = expr;
      if (expr.length > 0 && OPERATORS.includes(expr[0])) {
        body = expr.slice(1);
      }
      for (const spec of body.split(',')) {
        const name = spec;
        if (name) names.add(name);
      }
    }
    return [...names];
  }
}
